plugins/collector.go

```diff
@@ -12,46 +12,81 @@ type Collector interface {
 	Collect(b *metrics.Batch)
 }
 
-var collectors []Collector
+// Group collectors by their intervals to minimize number of long running
+// goroutines spawned.
+var collectors = make(map[time.Duration][]Collector)
 
-func AddCollector(c Collector) {
-	collectors = append(collectors, c)
+func AddCollector(c Collector, d time.Duration) {
+	collectors[d] = append(collectors[d], c)
 }
 
-func Collect(source, email, token string, duration time.Duration) {
-	client := librato.NewClient(email, token)
-	collect(client, source, collectors)
-	for _ = range time.Tick(duration) {
-		collect(client, source, collectors)
-	}
-}
-
-func collect(client *librato.Client, source string, collectors []Collector) {
-	start := time.Now()
+func Collect(source, email, token string, defaultDuration time.Duration, wg *sync.WaitGroup, stopSignal <-chan struct{}) {
+	var (
+		client = librato.NewClient(email, token)
+		stats  = metrics.NewBatch(source)
+		window = 500 * time.Millisecond
+	)
 
-	b := metrics.NewBatch(source)
-
-	// Do each collection in parallel. As a future improvement we might
-	// want to wait for up to a some time interval for collection to
-	// complete before publishing events to account for slow plugins.
-	var wg sync.WaitGroup
 	wg.Add(len(collectors))
-	for _, c := range collectors {
-		go func(collector Collector) {
-			collector.Collect(b)
-			wg.Done()
-		}(c)
+	// We spawn long running goroutines for all different invertals that will
+	// collect stats from the associated collectors. Goroutines exit cleanly
+	// when kill signal is sent.
+	for duration, plugins := range collectors {
+		if duration == 0 {
+			duration = defaultDuration
+		}
+		go collect(duration, plugins, stats, wg, stopSignal)
 	}
-	wg.Wait()
 
-	Logger.Printf("processed %d collectors in %v\n", len(collectors), time.Now().Sub(start))
+	// Flush data to librato as it becomes available.
+	go flusher(client, stats, window, wg, stopSignal)
+}
 
-	start = time.Now()
-	err := client.Send(b.Librato)
-	if err != nil {
-		Logger.Printf("send to librato failed: %v\n", err)
-		return
+func collect(d time.Duration, collectors []Collector, b *metrics.Batch, wg *sync.WaitGroup, stopSignal <-chan struct{}) {
+loop:
+	for {
+		select {
+		case <-stopSignal:
+			break loop
+		case <-time.Tick(d):
+			for _, c := range collectors {
+				go c.Collect(b)
+			}
+		}
+	}
+	wg.Done()
+}
+
+func flusher(client *librato.Client, b *metrics.Batch, window time.Duration, wg *sync.WaitGroup, stopSignal <-chan struct{}) {
+	// Stats have to be updated and remain unchanged for a specified window
+	// before we flush them out. The window is provide an oppurtunity to reduce
+	// number of flush updates when different intervals overlap.
+	var flushData = func() {
+		if data := b.Flush(window); data != nil {
```

Comment:
It would be less error prone to pull the batch at the configured collection intervals rather than this strategy.